msg_queue.go

```diff
@@ -8,17 +8,78 @@ import (
 // msgQueue defines the structure that holds message queues for different PBFT states
 type msgQueue struct {
 	// Heap implementation for the round change message queue
-	roundChangeStateQueue msgQueueImpl
+	roundChangeStateQueue *msgQueueImpl
 
 	// Heap implementation for the accept state message queue
-	acceptStateQueue msgQueueImpl
+	acceptStateQueue *msgQueueImpl
 
 	// Heap implementation for the validate state message queue
-	validateStateQueue msgQueueImpl
+	validateStateQueue *msgQueueImpl
+
+	pendingCommitStateQueue *msgQueueImpl
+
+	commitValidation *commitValidationRoutine
 
 	queueLock sync.Mutex
 }
 
+type commitValidationRoutine struct {
+	updateCh chan *RoundInfo2
+	closeCh  chan struct{}
+	queue    *msgQueueImpl
+}
+
+func (c *commitValidationRoutine) init() {
+	var round *RoundInfo2
+
+START:
+	for {
+		// read as many values as possible for the current view
+		val, _ := readValues(c.queue, round.view)
+		if val != nil {
+			// process in parallel
```

Comment:
We would need reference to state here (namely proposal hash, validator set and current view). Think that we would need a locking mechanism for the state, since we are accessing from multiple routines.